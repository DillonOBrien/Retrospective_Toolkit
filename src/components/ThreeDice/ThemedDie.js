/* eslint-disable no-console */
import React, { useEffect } from "react";
import { useThree } from "react-three-fiber";
import { useTexture } from "@react-three/drei";
import { useBox } from "@react-three/cannon";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";
import { useGesture, useDrag } from "react-use-gesture";
import { uniqueImageSet } from "../Dice/Dice";

const singleRollSound = () => {
  new Audio("/diceRoll.m4a").play();
};

/* istanbul ignore next */
const ThemedDie = (props) => {
  const {
    diceInitPos,
    rerollAllToggle,
    rerollValue,
    rerollDieToggle,
    imageSet,
    setImages,
    geom,
    setOrbitControl,
  } = props;
  // const [pos, setPos] = useState(diePos);
  const textures = useTexture([...imageSet]);

  const [mesh, api] = useBox(() => ({
    mass: 300,
    inertia: 13,
    rotation: [
      Math.random() * Math.PI,
      Math.random() * Math.PI,
      Math.random() * Math.PI,
    ],
    linearDamping: 0.5,
    angularDamping: 0.1,
    material: { restitution: 0.3 },
  }));

  const { size, viewport } = useThree();
  const aspect = size.width / viewport.width;
  const isDragging = React.useRef(false);

  const clickBind = useGesture({
    onClick: () => {
      if (isDragging.current) return;
      rerollDieToggle(!rerollValue);
      singleRollSound();
    },
  });

  const dragBind = useDrag(
    ({ offset: [x, y], first, last }) => {
      if (first) {
        setOrbitControl(false);
        isDragging.current = true;
      } else if (last) {
        isDragging.current = false;
        setOrbitControl(true);
      }
      api.position.set(x / aspect, 1.5, y / aspect);
    },
    {
      threshold: 1,
      delay: 500,
    }
  );

  useEffect(() => {
    api.position.set(diceInitPos[0], diceInitPos[1], diceInitPos[2]);
    api.velocity.set(15, 0, -10);
    api.angularVelocity.set(-15, 2, -10);
    setImages(uniqueImageSet);
  }, [
    api.angularVelocity,
    api.position,
    api.velocity,
    diceInitPos,
    rerollAllToggle,
    rerollValue,
    setImages,
  ]);

  return (
    <mesh
      ref={mesh}
      geometry={geom}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...dragBind()}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...clickBind()}
    >
      <boxBufferGeometry />
      {textures.map((image) => (
        <meshStandardMaterial
          key={uuidv4()}
          flatShading
          roughness={0.8}
          attachArray="material"
          map={image}
        />
      ))}
    </mesh>
  );
};

ThemedDie.propTypes = {
  diceInitPos: PropTypes.arrayOf(PropTypes.number).isRequired,
  imageSet: PropTypes.arrayOf(PropTypes.string).isRequired,
  setImages: PropTypes.func.isRequired,
  rerollAllToggle: PropTypes.bool.isRequired,
  rerollValue: PropTypes.bool.isRequired,
  rerollDieToggle: PropTypes.func.isRequired,
  geom: PropTypes.shape().isRequired,
  setOrbitControl: PropTypes.func.isRequired,
};

export default ThemedDie;
