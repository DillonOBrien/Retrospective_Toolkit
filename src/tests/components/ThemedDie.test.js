import React, { Suspense } from "react";
import {
  render,
  screen,
  cleanup,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { Canvas } from "react-three-fiber";
import { Physics } from "@react-three/cannon";
import ThemedDie from "../../components/ThreeDice/ThemedDie";

afterEach(cleanup);

it("should do a snapshot test on ThemedDie", async () => {
  render(
    <Canvas concurrent style={{ width: "100vw", height: "500px" }}>
      <Physics>
        <Suspense fallback={<div>Loading...</div>}>
          <ThemedDie
            theme="random"
            dicePos={[-13, 5, 6]}
            rerollToggle={false}
          />
        </Suspense>
      </Physics>
    </Canvas>
  );
  const loading = () => screen.getByText("Loading...");
  await waitForElementToBeRemoved(loading);
  screen.debug();
  // expect(container).toMatchSnapshot();
});
