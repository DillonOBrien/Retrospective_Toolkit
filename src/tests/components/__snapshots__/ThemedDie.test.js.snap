// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should do a snapshot test on ThemedDie 1`] = `
<Suspense
  fallback={
    <div>
      Loading...
    </div>
  }
>
  <ThemedDie
    dicePos={
      Array [
        -13,
        5,
        6,
      ]
    }
    rerollToggle={false}
    testID="testDie"
    theme="random"
  />
</Suspense>
`;
